README.md
# LTI 1.3 Demo Application

[![Latest Version](https://img.shields.io/github/tag/oat-sa/demo-lti1p3.svg?style=flat&label=release)](https://github.com/oat-sa/demo-lti1p3/tags)
[![License GPL2](http://img.shields.io/badge/licence-GPL%202.0-blue.svg)](http://www.gnu.org/licenses/gpl-2.0.html)

> [Symfony](https://symfony.com/) demo application for LTI 1.3, to act as a [LTI platform and / or tool](http://www.imsglobal.org/spec/lti/v1p3/#platforms-and-tools-0).

This demo application based on the following packages:
- [LTI 1.3 symfony bundle](https://github.com/oat-sa/bundle-lti1p3)
- [LTI 1.3 NRPS library](https://github.com/oat-sa/lib-lti1p3-nrps)
- [LTI 1.3 deep linking library](https://github.com/oat-sa/lib-lti1p3-deep-linking)
- [LTI 1.3 basic outcome library](https://github.com/oat-sa/lib-lti1p3-basic-outcome)
- [LTI 1.3 core library](https://github.com/oat-sa/lib-lti1p3-core)


## Table of Contents

- [Live demo](#live-demo)
- [Specifications](#specifications)
- [Installation](#installation)
- [Configuration](#configuration)
- [Usage](#usage)

## Live demo
 
To try it live: [https://lti.showcase.gcp.taocloud.org](https://lti.showcase.gcp.taocloud.org)

## Specifications
 
This demo application respect the following LTI 1.3 [IMS](http://www.imsglobal.org) specifications:
- [IMS Security](https://www.imsglobal.org/spec/security/v1p0)
- [IMS LTI 1.3 NRPS](https://www.imsglobal.org/spec/lti-nrps/v2p0)
- [IMS LTI 1.3 Deep Linking](https://www.imsglobal.org/spec/lti-dl/v2p0/)
- [IMS LTI 1.3 Basic Outcome](https://www.imsglobal.org/spec/lti-bo/v1p1)
- [IMS LTI 1.3 Core](http://www.imsglobal.org/spec/lti/v1p3)

## Installation

You need to ensure first you have the following installed:
- [docker](https://docs.docker.com/get-docker/)
- [docker-compose](https://docs.docker.com/compose/install/)

After cloning this repository, you can build the [provided docker stack](docker-compose.yml):
```console
$ docker-compose up -d
```

Then, install required dependencies with [composer](https://hub.docker.com/_/composer):
```console
$ docker run --rm --interactive --tty \
  --volume $PWD:/app \
  composer install
```

For Windows users:
- you may have to do `--volume %cd%:/app` instead
- with powershell, you may have to do `--volume ${PWD}:/app` instead

## Configuration

### Platforms, tools and registrations

Since this demo application relies on [LTI 1.3 symfony bundle](https://github.com/oat-sa/bundle-lti1p3), you can find [here](https://github.com/oat-sa/bundle-lti1p3/blob/master/doc/quickstart/configuration.md) instructions to configure it.

### Customization

You can find in the [config/demo](config/demo) folder configuration files to customize the LTI 1.3 demo application:
- [claims.yaml](config/demo/claims.yaml): configurable editor claims list
- [deep_linking.yaml](config/demo/deep_linking.yaml): configurable deep linking resources list
- [users.yaml](config/demo/users.yaml): configurable users list

## Usage

### Application

After installation, the LTI 1.3 demo application is available on [http://localhost:8888](http://localhost:8888)

### Services

After installation, the following services are available:

| Name                                 | Description                      |
|--------------------------------------|----------------------------------|
| demo_lti1p3_nginx                    | application nginx web server     |
| demo_lti1p3_phpfpm                   | application php-fpm              |
| demo_lti1p3_redis                    | application cache                |
| demo_lti1p3_redis_commander          | application cache administration |

You can access:

| Name                                 | URL                                            |
|--------------------------------------|------------------------------------------------|
| demo_lti1p3_nginx                    | [http://localhost:8888](http://localhost:8888) |
| demo_lti1p3_redis_commander          | [http://localhost:8081](http://localhost:8081) |

### Tests

To run tests:

```console
$ bin/phpunit
```
**Note**: see [phpunit.xml.dist](phpunit.xml.dist) for available test suites.


CHANGELOG.md
CHANGELOG
=========

1.9.0
-----

* Added tool launch tab for AGS
* Added platform CRUD for AGS  
* Updated NRPS and Basic Outcome service endpoints to use new request handler system
* Updated library and bundle dependencies to 5.x

1.8.0
-----

* Added platform basic outcome list
* Updated platform NRPS memberships list

1.7.0
-----

* Added default NRPS membership (based on users configuration file)
* Added platform CRUD for custom NRPS memberships (cache storage)
* Updated default NRPS service url claim editor to match default NRPS membership
* Updated context identifier claim editor to default

1.6.0
-----

* Added possibility to provide custom user on launch (for resource link and deep linking)

1.5.0
-----

* Added possibility to configure resource link at launch (fallback to auto generated one when not specified)
* Added resource link claim in claims editor

1.4.0
-----

* Added PHP 8 support (and kept >=7.2)
* Added version indicator on homepage
* Updated all LTI dependencies (for PHP 8 support)
* Updated demo app code to match new LTI dependencies breaking changes

1.3.0
-----

* Added claims editor JSON validation (lti resource link, deep link)
* Added possibility to share generator forms via url (lti resource link, deep link, tool service client)

1.2.0
-----

* Added more details to tool generic service client response interface

1.1.0
-----

* Added launch url editors for LTI launch and deep linking
* Reworked interface for services callers

1.0.0
-----

* Added platform side message features (LTI resource link launch, deep linking launch)
* Added platform side service features (NRPS server, basic outcome server)
* Added tool side message features (LTI resource link launch, deep linking content selection)
* Added tool side service features (NRPS client, basic outcome client, generic service client)
* Added docker & k8s compliance
* Added documentation


tests/Unit/TraitsTest.php
<?php

namespace App\Tests\Unit;

use OAT\Library\Lti1p3Core\Tests\Traits\DomainTestingTrait;
use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;

class TraitsTest extends KernelTestCase
{
    use DomainTestingTrait;

    /**
     * Test to ensure core lib testing traits can be used into an SF application
     */
    public function testStuff(): void
    {
        $registration = $this->createTestRegistration();

        $token = $this->buildJwt(
            [],
            [
                'some' => 'value'
            ],
            $registration->getPlatformKeyChain()->getPrivateKey()
        );

        $parsedToken = $this->parseJwt($token->toString());

        $this->assertTrue($this->verifyJwt($parsedToken, $registration->getPlatformKeyChain()->getPublicKey()));
        $this->assertEquals('value', $parsedToken->getClaims()->get('some'));
    }
}


src/Ags/LineItemRepository.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace App\Ags;

use OAT\Library\Lti1p3Ags\Model\LineItem\LineItemCollection;
use OAT\Library\Lti1p3Ags\Model\LineItem\LineItemCollectionInterface;
use OAT\Library\Lti1p3Ags\Model\LineItem\LineItemInterface;
use OAT\Library\Lti1p3Ags\Repository\LineItemRepositoryInterface;
use Psr\Cache\CacheItemPoolInterface;

class LineItemRepository implements LineItemRepositoryInterface
{
    private const CACHE_KEY = 'lti1p3-ags-line-items';

    /** @var CacheItemPoolInterface */
    private $cache;

    public function __construct(CacheItemPoolInterface $cache)
    {
        $this->cache = $cache;
    }

    public function find(string $lineItemIdentifier): ?LineItemInterface
    {
        $cache = $this->cache->getItem(self::CACHE_KEY);

        if ($cache->isHit()) {
            $lineItems = $cache->get();

            if (array_key_exists($lineItemIdentifier, $lineItems)) {
                return $lineItems[$lineItemIdentifier];
            }
        }

        return null;
    }

    public function findBy(
        ?string $contextIdentifier = null,
        ?string $resourceIdentifier = null,
        ?string $resourceLinkIdentifier = null,
        ?string $tag = null,
        ?int $limit = null,
        ?int $offset = null
    ): LineItemCollectionInterface {
        $cache = $this->cache->getItem(self::CACHE_KEY);

        $foundLineItems = [];
        $hasNext = false;

        if ($cache->isHit()) {

            $lineItems = $cache->get();

            foreach ($lineItems as $lineItem) {
                $found = true;

                if (null !== $contextIdentifier) {
                    $found = $found && $lineItem->getContextIdentifier() === $contextIdentifier;
                }

                if (null !== $resourceIdentifier) {
                    $found = $found && $lineItem->getResourceIdentifier() === $resourceIdentifier;
                }

                if (null !== $resourceLinkIdentifier) {
                    $found = $found && $lineItem->getResourceLinkIdentifier() === $resourceLinkIdentifier;
                }

                if (null !== $tag) {
                    $found = $found && $lineItem->getTag() === $tag;
                }

                if ($found) {
                    $foundLineItems[] = $lineItem;
                }
            }

            $hasNext = ($limit ? intval($limit) : 0) >= sizeof($lineItems);
        }

        return new LineItemCollection(
            array_slice($foundLineItems, $offset ? intval($offset) : 0, $limit ? intval($limit) : null),
            $hasNext
        );
    }

    public function save(LineItemInterface $lineItem): LineItemInterface
    {
        $cache = $this->cache->getItem(self::CACHE_KEY);

        $lineItems = $cache->get();

        $lineItems[$lineItems->getIdentifier()] = $lineItems;

        $cache->set($lineItems);

        $this->cache->save($cache);
    }

    public function delete(string $lineItemIdentifier): void
    {
        $cache = $this->cache->getItem(self::CACHE_KEY);

        $lineItems = $cache->get();

        unset($lineItems[$lineItems->getIdentifier()]);

        $cache->set($lineItems);

        $this->cache->save($cache);
    }
}


src/BasicOutcome/BasicOutcomeProcessor.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace App\BasicOutcome;

use OAT\Library\Lti1p3BasicOutcome\Service\Server\Processor\BasicOutcomeServiceServerProcessorInterface;
use OAT\Library\Lti1p3BasicOutcome\Service\Server\Processor\Result\BasicOutcomeServiceServerProcessorResult;
use OAT\Library\Lti1p3Core\Registration\RegistrationInterface;
use Psr\Cache\CacheItemPoolInterface;

class BasicOutcomeProcessor implements BasicOutcomeServiceServerProcessorInterface
{
    public const CACHE_KEY = 'lti1p3-nrps-basic-outcomes';

    /** @var CacheItemPoolInterface */
    private $cache;

    public function __construct(CacheItemPoolInterface $cache)
    {
        $this->cache = $cache;
    }

    public function processReadResult(
        RegistrationInterface $registration,
        string $sourcedId
    ): BasicOutcomeServiceServerProcessorResult {

        $basicOutcomeCache = $this->cache->getItem(self::CACHE_KEY);

        if ($basicOutcomeCache->isHit()) {

            $basicOutcomeList = $basicOutcomeCache->get();
            $basicOutcomeKey = $this->getBasicOutcomeKey($registration, $sourcedId);

            if (array_key_exists($basicOutcomeKey, $basicOutcomeList)) {

                $score = floatval($basicOutcomeList[$basicOutcomeKey]['score']);
                $language = $basicOutcomeList[$basicOutcomeKey]['language'] ?? null;

                return new BasicOutcomeServiceServerProcessorResult(
                    true,
                    sprintf('The current score for sourced id %s is %s and for language %s', $sourcedId, $score, $language),
                    $score,
                    $language
                );
            }

            return new BasicOutcomeServiceServerProcessorResult(
                true,
                sprintf('The current score for sourced id %s is non existing', $sourcedId)
            );
        }

        return new BasicOutcomeServiceServerProcessorResult(
            true,
            sprintf('The current score for sourced id %s is non existing', $sourcedId)
        );
    }

    public function processReplaceResult(
        RegistrationInterface  $registration,
        string $sourcedId,
        float $score,
        string $language = 'en'
    ): BasicOutcomeServiceServerProcessorResult {

        $basicOutcomeCache = $this->cache->getItem(self::CACHE_KEY);

        $basicOutcomeList = [];

        if ($basicOutcomeCache->isHit()) {
            $basicOutcomeList = $basicOutcomeCache->get();
        }

        $basicOutcomeKey = $this->getBasicOutcomeKey($registration, $sourcedId);

        $basicOutcomeList[$basicOutcomeKey] = [
            'score' => $score,
            'language' => $language,
            'registration' => $registration->getIdentifier(),
            'tool' => $registration->getTool()->getIdentifier(),
        ];

        $basicOutcomeCache->set($basicOutcomeList);

        $this->cache->save($basicOutcomeCache);

        return new BasicOutcomeServiceServerProcessorResult(
            true,
            sprintf('The new score for sourced id %s is now %s and for language %s', $sourcedId, $score, $language),
            $score,
            $language
        );
    }

    public function processDeleteResult(
        RegistrationInterface $registration,
        string $sourcedId
    ): BasicOutcomeServiceServerProcessorResult {

        $basicOutcomeCache = $this->cache->getItem(self::CACHE_KEY);

        $basicOutcomeKey = $this->getBasicOutcomeKey($registration, $sourcedId);

        $basicOutcomeList = [];

        if ($basicOutcomeCache->isHit()) {
            $basicOutcomeList = $basicOutcomeCache->get();
        }

        unset($basicOutcomeList[$basicOutcomeKey]);

        $basicOutcomeCache->set($basicOutcomeList);

        $this->cache->save($basicOutcomeCache);

        return new BasicOutcomeServiceServerProcessorResult(
            true,
            sprintf('The score for sourced id %s has been deleted', $sourcedId)
        );
    }

    private function getBasicOutcomeKey(RegistrationInterface $registration, string $sourcedId): string
    {
        return sprintf('%s-%s', $registration->getIdentifier(), $sourcedId);
    }
}


src/Action/Tool/Ajax/Nrps/NrpsServiceClientAction.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace App\Action\Tool\Ajax\Nrps;

use OAT\Library\Lti1p3Core\Registration\RegistrationRepositoryInterface;
use OAT\Library\Lti1p3Nrps\Service\Client\MembershipServiceClient;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Twig\Environment;

class NrpsServiceClientAction
{
    /** @var Environment */
    private $twig;

    /** @var MembershipServiceClient */
    private $client;

    /** @var RegistrationRepositoryInterface */
    private $repository;

    public function __construct(
        Environment $twig,
        MembershipServiceClient $client,
        RegistrationRepositoryInterface $repository
    ) {
        $this->twig = $twig;
        $this->client = $client;
        $this->repository = $repository;
    }

    public function __invoke(Request $request): Response
    {
        $membership = $this->client->getContextMembership(
            $this->repository->find($request->get('registration')),
            $request->get('url'),
            $request->get('role'),
            intval($request->get('limit'))
        );

        return new Response(
            $this->twig->render('tool/ajax/nrps.html.twig', ['membership' => $membership])
        );
    }
}


src/Action/Platform/Service/Nrps/MembershipServiceAction.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace App\Action\Platform\Service\Nrps;

use OAT\Bundle\Lti1p3Bundle\Security\Authentication\Token\Service\LtiServiceSecurityToken;
use OAT\Library\Lti1p3Nrps\Service\Server\Handler\MembershipServiceServerRequestHandler;
use Symfony\Bridge\PsrHttpMessage\HttpFoundationFactoryInterface;
use Symfony\Bridge\PsrHttpMessage\HttpMessageFactoryInterface;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Security\Core\Security;

class MembershipServiceAction
{
    /** @var MembershipServiceServerRequestHandler */
    private $handler;

    /** @var Security */
    private $security;

    /** @var HttpFoundationFactoryInterface */
    private $httpFoundationFactory;

    /** @var HttpMessageFactoryInterface */
    private $psr7Factory;

    public function __construct(
        MembershipServiceServerRequestHandler $handler,
        Security $security,
        HttpFoundationFactoryInterface $httpFoundationFactory,
        HttpMessageFactoryInterface $psr7Factory
    ) {
        $this->handler = $handler;
        $this->security = $security;
        $this->httpFoundationFactory = $httpFoundationFactory;
        $this->psr7Factory = $psr7Factory;
    }

    public function __invoke(Request $request): Response
    {
        /** @var LtiServiceSecurityToken $token */
        $token = $this->security->getToken();

        $basicOutcomeResponse = $this->handler->handleServiceRequest(
            $token->getRegistration(),
            $this->psr7Factory->createRequest($request)
        );

        return $this->httpFoundationFactory->createResponse($basicOutcomeResponse);
    }
}


src/Action/Platform/Service/Ags/LineItem/ListLineItemsAction.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace App\Action\Platform\Service\Ags\LineItem;

use OAT\Bundle\Lti1p3Bundle\Security\Authentication\Token\Service\LtiServiceSecurityToken;
use OAT\Library\Lti1p3Ags\Service\LineItem\Server\Handler\ListLineItemServiceServerRequestHandler;
use Symfony\Bridge\PsrHttpMessage\HttpFoundationFactoryInterface;
use Symfony\Bridge\PsrHttpMessage\HttpMessageFactoryInterface;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Security\Core\Security;

class ListLineItemsAction
{
    /** @var ListLineItemServiceServerRequestHandler */
    private $handler;

    /** @var Security */
    private $security;

    /** @var HttpFoundationFactoryInterface */
    private $httpFoundationFactory;

    /** @var HttpMessageFactoryInterface */
    private $psr7Factory;

    public function __construct(
        ListLineItemServiceServerRequestHandler $handler,
        Security $security,
        HttpFoundationFactoryInterface $httpFoundationFactory,
        HttpMessageFactoryInterface $psr7Factory
    ) {
        $this->handler = $handler;
        $this->security = $security;
        $this->httpFoundationFactory = $httpFoundationFactory;
        $this->psr7Factory = $psr7Factory;
    }

    public function __invoke(Request $request): Response
    {
        /** @var LtiServiceSecurityToken $token */
        $token = $this->security->getToken();

        $basicOutcomeResponse = $this->handler->handleServiceRequest(
            $token->getRegistration(),
            $this->psr7Factory->createRequest($request)
        );

        return $this->httpFoundationFactory->createResponse($basicOutcomeResponse);
    }
}


src/Action/Tool/Ajax/BasicOutcome/BasicOutcomeServiceClientAction.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace App\Action\Tool\Ajax\BasicOutcome;

use OAT\Library\Lti1p3BasicOutcome\Message\BasicOutcomeMessageInterface;
use OAT\Library\Lti1p3BasicOutcome\Service\Client\BasicOutcomeServiceClient;
use OAT\Library\Lti1p3Core\Registration\RegistrationRepositoryInterface;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Twig\Environment;

class BasicOutcomeServiceClientAction
{
    /** @var Environment */
    private $twig;

    /** @var BasicOutcomeServiceClient */
    private $client;

    /** @var RegistrationRepositoryInterface */
    private $repository;

    public function __construct(
        Environment $twig,
        BasicOutcomeServiceClient $client,
        RegistrationRepositoryInterface $repository
    ) {
        $this->twig = $twig;
        $this->client = $client;
        $this->repository = $repository;
    }

    public function __invoke(Request $request): Response
    {
        switch ($request->get('operation')) {
            case BasicOutcomeMessageInterface::TYPE_READ_RESULT:
                $basicOutcomeResponse = $this->client->readResult(
                    $this->repository->find($request->get('registration')),
                    $request->get('url'),
                    $request->get('resultSourcedId')
                );
                break;
            case BasicOutcomeMessageInterface::TYPE_REPLACE_RESULT:
                $basicOutcomeResponse = $this->client->replaceResult(
                    $this->repository->find($request->get('registration')),
                    $request->get('url'),
                    $request->get('resultSourcedId'),
                    (float)$request->get('score'),
                    $request->get('language')
                );
                break;
            case BasicOutcomeMessageInterface::TYPE_DELETE_RESULT:
                $basicOutcomeResponse = $this->client->deleteResult(
                    $this->repository->find($request->get('registration')),
                    $request->get('url'),
                    $request->get('resultSourcedId')
                );
                break;
        }

        return new Response(
            $this->twig->render(
                'tool/ajax/basic-outcome.html.twig',
                [
                    'response' => $basicOutcomeResponse,
                ]
            )
        );
    }
}


src/Nrps/MembershipServiceServerBuilder.php
<?php

/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
 */

declare(strict_types=1);

namespace App\Nrps;

use OAT\Library\Lti1p3Core\Registration\RegistrationInterface;
use OAT\Library\Lti1p3Nrps\Model\Member\MemberCollection;
use OAT\Library\Lti1p3Nrps\Model\Member\MemberInterface;
use OAT\Library\Lti1p3Nrps\Model\Membership\MembershipInterface;
use OAT\Library\Lti1p3Nrps\Service\Server\Builder\MembershipServiceServerBuilderInterface;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\RequestStack;
use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;

class MembershipServiceServerBuilder implements MembershipServiceServerBuilderInterface
{
    /** @var RequestStack */
    private $requestStack;

    /** @var MembershipRepository */
    private $repository;

    /** @var DefaultMembershipFactory */
    private $factory;

    public function __construct(
        RequestStack $requestStack,
        MembershipRepository $repository,
        DefaultMembershipFactory $factory
    ) {
        $this->requestStack = $requestStack;
        $this->repository = $repository;
        $this->factory = $factory;
    }

    public function buildResourceLinkMembership(
        RegistrationInterface $registration,
        string $resourceLinkIdentifier,
        string $role = null,
        int $limit = null,
        int $offset = null
    ): MembershipInterface {
        return $this->build($registration, $role, $limit, $offset);
    }

    public function buildContextMembership(
        RegistrationInterface $registration,
        string $role = null,
        int $limit = null,
        int $offset = null
    ): MembershipInterface {
        return $this->build($registration, $role, $limit, $offset);
    }

    private function build(
        RegistrationInterface $registration,
        string $role = null,
        int $limit = null,
        int $offset = null
    ): MembershipInterface {

        $request = $this->requestStack->getCurrentRequest();
        $routeParameters = $request->attributes->get('_route_params');

        $membershipIdentifier = $routeParameters['membershipIdentifier'] ?? null;
        $contextIdentifier = $routeParameters['contextIdentifier'] ?? null;

        if ($membershipIdentifier === 'default') {
            $membership = $this->factory->create();
        } else {
            $membership = $this->repository->find($membershipIdentifier);
        }

        if (null === $membership || $membership->getContext()->getIdentifier() !== $contextIdentifier) {
            throw new NotFoundHttpException(
                sprintf('Membership with context %s and identifier %s cannot be found', $contextIdentifier, $membershipIdentifier)
            );
        }

        $filteredMembers = array_filter(
            $membership->getMembers()->all(),
            static function (MemberInterface $member) use ($role) {
                if (null === $role) {
                    return true;
                }

                return in_array($role, $member->getRoles());
            }
        );

        return $membership
            ->setMembers(new MemberCollection(array_slice($filteredMembers, $offset ?? 0, $limit ?? null)))
            ->setRelationLink($this->buildRelationLink($request, sizeof($filteredMembers), $role, $limit, $offset));
    }

    private function buildRelationLink(
        Request $request,
        int $totalCount,
        string $role = null,
        int $limit = null,
        int $offset = null
    ): ?string {
        if ((($limit ?? 0) + ($offset ?? 0)) < $totalCount) {
            $parsedUrl = parse_url(urldecode($request->getUri()));
            parse_str($parsedUrl['query'] ?? '', $parsedQuery);

            $nextOffset = $limit ?? 0;

            if ($parsedQuery['offset'] ?? false) {
                $nextOffset += $parsedQuery['offset'];
            }

            $relUrl = sprintf(
                '%s://%s%s%s',
                $parsedUrl['scheme'],
                $parsedUrl['host'],
                $parsedUrl['port'] ?? false ? ':' . $parsedUrl['port'] : '',
                $parsedUrl['path']
            );

            $relUrlQuery = array_filter(
                [
                    'role' => $role,
                    'offset' => $nextOffset
                ]
            );

            return $relUrl . '?' . http_build_query($relUrlQuery) . '; rel="next"';
        }

        return null;
    }
}
